File: RestClient/src/com/kalicinscy/web/restclient/client/JSONViewer.java
package com.kalicinscy.web.restclient.client;

import java.util.Iterator;
import java.util.Set;

import com.google.gwt.json.client.JSONArray;
import com.google.gwt.json.client.JSONBoolean;
import com.google.gwt.json.client.JSONNull;
import com.google.gwt.json.client.JSONNumber;
import com.google.gwt.json.client.JSONObject;
import com.google.gwt.json.client.JSONParser;
import com.google.gwt.json.client.JSONString;
import com.google.gwt.json.client.JSONValue;
import com.google.gwt.user.client.ui.HTML;

public class JSONViewer {

	JSONValue jsonValue = null;
	private final int INDENT_VALUE = 15;

	public JSONViewer(String data) {
		try {
			this.jsonValue = JSONParser.parseLenient(data);
		} catch (NullPointerException e) {

		} catch (IllegalArgumentException e) {

		}
	}

	public void setHTML(HTML body) {
		if (this.jsonValue == null) {
			return;
		}
		String html = this.parse( this.jsonValue, 1 );
		
		String data = "<div class=\"pretty-print\">";
		data += html;
		data += "</div>";
		
		body.setHTML(data);
	}
	
	private String parse(JSONValue data, int indent){
		String result = "";
		JSONObject obj = data.isObject(); //JSON Object or null
		JSONArray arr = data.isArray(); //JSON array or null
		JSONString str = data.isString();//JSON string or null
		JSONBoolean bool = data.isBoolean();//JSON boolean or null
		JSONNull isNull = data.isNull(); //JSON null or null
		JSONNumber number = data.isNumber();//JSON number ornull
		
		
		
		if( obj != null ){
			result += this.parseObject(obj, indent);
		} else if( arr != null ) {
			result += this.parseArray(arr, indent);
		} else if( str != null ) {
			result += this.parseStringValue(str);
		} else if( bool != null ) {
			result += this.parseBooleanValue(bool);
		} else if( isNull != null ) {
			result += this.parseNullValue(isNull);
		} else if( number != null ) {
			result += this.parseNumericValue(number);
		}
		
		
		return result;
	}
	
	private String parseNumericValue(JSONNumber number) {
		String result = "";
		result += "<span class=\"json-numeric\">";
		result += number.toString();
		result += "</span>";
		return result;
	}

	private String parseNullValue(JSONNull isNull) {
		String result = "";
		result += "<span class=\"json-null\">";
		result += "null";
		result += "</span>";
		return result;
	}

	private String parseBooleanValue(JSONBoolean bool) {
		String result = "";
		result += "<span class=\"json-boolean\">";
		if( bool != null )
			result += String.valueOf(bool.booleanValue());
		else 
			result += "null";
		result += "</span>";
		return result;
	}

	private String parseStringValue(JSONString str) {
		String result = "";
		result += "<span class=\"json-punctuation\">&quot;</span>";
		result += "<span class=\"json-string\">";
		if( str != null )
			result += str.stringValue();
		else
			result += "null";
		result += "</span>";
		result += "<span class=\"json-punctuation\">&quot;</span>";
		return result;
	}

	private String parseObject(JSONObject object, int indent) {
		String result = "";
		
		result += "<span class=\"json-punctuation\">{</span>";
		
		Set<String> keys = object.keySet();
		Iterator<String> it = keys.iterator();
		
		while( it.hasNext() ){
			int indentValue = indent*INDENT_VALUE;
			
			result += "<div style=\"text-indent:"+indentValue+"px\" class=\"node\">";
			String key = it.next();
			JSONValue value = object.get(key);
			String data = this.parse(value, indent+1);
			result += this.parseKey(key) +": "+data;
			if(it.hasNext()){
				result += "<span class=\"json-punctuation\">,</span>";
			}
			result += "</div>";
		}
		
		
		result += "<span class=\"json-punctuation\">}</span>";
		return result;
	}
	
	
	
	
	private String parseArray(JSONArray array, int indent) {
		String result = "";
		
		result += "<span class=\"json-punctuation\">[</span>";
		int cnt = array.size();
		
		result += "<span class=\"json-array-counter\">("+cnt+")</span>";
		
		for( int i = 0; i<cnt; i++ ){
			int indentValue = indent*INDENT_VALUE;
			
			result += "<div style=\"text-indent:"+indentValue+"px\" class=\"node\">";
			JSONValue value = array.get(i);
			result += this.parse( value, indent+1 );
			if( i<cnt-1 ){
				result += "<span class=\"json-punctuation\">,</span>";
			}
			result += "</div>";
		}
		
		result += "<span class=\"json-punctuation\">]</span>";
		return result;
	}
	
	private String parseKey( String key ){
		String result = "";
		result += "<span class=\"json-punctuation\">&quot;</span>";
		result += "<span class=\"json-key-name\">"+key+"</span>";
		result += "<span class=\"json-punctuation\">&quot;</span>";
		return result;
	}
	
}

File: app/scripts/workers/jsonviewer.js
'use strict';
/* global self */

var SafeHtmlUtils = {
  AMP_RE: new RegExp(/&/g),
  GT_RE: new RegExp(/>/g),
  LT_RE: new RegExp(/</g),
  SQUOT_RE: new RegExp(/'/g),
  QUOT_RE: new RegExp(/"/g),

  htmlEscape: function(s) {
    if (s.indexOf('&') !== -1) {
      s = s.replace(SafeHtmlUtils.AMP_RE, '&amp;');
    }
    if (s.indexOf('<') !== -1) {
      s = s.replace(SafeHtmlUtils.LT_RE, '&lt;');
    }
    if (s.indexOf('>') !== -1) {
      s = s.replace(SafeHtmlUtils.GT_RE, '&gt;');
    }
    if (s.indexOf('"') !== -1) {
      s = s.replace(SafeHtmlUtils.QUOT_RE, '&quot;');
    }
    if (s.indexOf('\'') !== -1) {
      s = s.replace(SafeHtmlUtils.SQUOT_RE, '&#39;');
    }
    return s;
  }
};

class JSONViewer {
  constructor(data) {
    var jsonData = data.json;
    this.rawData = data.raw || '';
    this._numberIndexes = {}; // Regexp number indexes
    this.linkRegExp = /([^"\s&;<>]*:\/\/[^"\s<>]*)(&quot;|&lt;|&gt;)?/gim;
    this.jsonValue = null;
    this.latestError = null;
    this.elementsCounter = 0;
    if (typeof jsonData === 'string') {
      try {
        this.jsonValue = JSON.parse(jsonData);
      } catch (e) {
        this.latestError = e.message;
      }
    } else {
      this.jsonValue = jsonData;
    }
  }
  /**
   * Get created HTML content.
   */
  getHTML() {
    var parsedData = '<div class="prettyPrint">';
    parsedData += this.parse(this.jsonValue);
    parsedData += '</div>';

    var replace = '<a title="Click to insert into URL field" response-anchor href="$1">$1</a>';
    var match = parsedData.match(this.linkRegExp);
    replace = replace.replace(/\$0/, match);
    parsedData = parsedData.replace(this.linkRegExp, replace);

    return parsedData;
  }
  /**
   * Parse JSON data
   */
  parse(data, opts) {
    opts = opts || {};

    var result = '';
    if (data === null) {
      result += this.parseNullValue();
    } else if (typeof data === 'number') {
      result += this.parseNumericValue(data);
    } else if (typeof data === 'boolean') {
      result += this.parseBooleanValue(data);
    } else if (typeof data === 'string') {
      result += this.parseStringValue(data);
    } else if (data instanceof Array) {
      result += this.parseArray(data);
    } else {
      result += this.parseObject(data);
    }
    if (opts.hasNextSibling && !opts.holdComa) {
      result += '<span class="punctuation hidden">,</span>';
    }
    return result;
  }

  parseNullValue() {
    var result = '';
    result += '<span class="nullValue">';
    result += 'null';
    result += '</span>';
    return result;
  }

  parseNumericValue(number) {
    var expectedNumber;
    if (number > Number.MAX_SAFE_INTEGER) {
      let comp = String(number);
      comp = comp.substr(0, 15);
      let r = new RegExp(comp + '(\\d+),', 'gim');
      if (comp in this._numberIndexes) {
        r.lastIndex = this._numberIndexes[comp];
      }
      let result = r.exec(this.rawData);
      if (result) {
        this._numberIndexes[comp] = result.index;
        expectedNumber = comp + result[1];
      }
    }

    var result = '';
    result += '<span class="numeric">';
    if (expectedNumber) {
      result += '<js-max-number-error class="number-error" expected-number="' + expectedNumber +
        '">';
    }
    result += number + '';
    if (expectedNumber) {
      result += '</js-max-number-error>';
    }
    result += '</span>';
    return result;
  }

  parseBooleanValue(bool) {
    var result = '';
    result += '<span class="booleanValue">';
    if (bool !== null && bool !== undefined) {
      result += bool + '';
    } else {
      result += 'null';
    }
    result += '</span>';
    return result;
  }

  parseStringValue(str) {
    var result = '';
    var value = str || '';
    if (value !== null && value !== undefined) {
      value = SafeHtmlUtils.htmlEscape(value);
      if (value.slice(0, 1) === '/') {
        value = '<a title="Click to insert into URL field" response-anchor add-root-url href="' +
          value + '">' + value + '</a>';
      }
    } else {
      value = 'null';
    }
    result += '<span class="punctuation">&quot;</span>';
    result += '<span class="stringValue">';
    result += value;
    result += '</span>';
    result += '<span class="punctuation">&quot;</span>';
    return result;
  }

  parseObject(object) {

    var result = '';
    result += '<div class="punctuation brace">{</div>';
    result += '<div collapse-indicator class="info-row">...</div>';

    Object.getOwnPropertyNames(object).forEach((key, i, arr) => {
      let value = object[key];

      let lastSibling = (i + 1) === arr.length;
      let parseOpts = {
        hasNextSibling: !lastSibling
      };
      if (value instanceof Array) {
        parseOpts.holdComa = true;
      }
      var elementNo = this.elementsCounter++;
      var data = this.parse(value, parseOpts);
      var hasManyChildren = this.elementsCounter - elementNo > 1;
      result += '<div data-element="' + elementNo + '" style="margin-left: 24px" class="node">';
      var _nan = isNaN(key);
      if (_nan) {
        result += '"';
      }
      result += this.parseKey(key);
      if (_nan) {
        result += '"';
      }
      result += ': ' + data;
      if (hasManyChildren) {
        result += '<div data-toggle="' + elementNo + '" class="rootElementToggleButton"></div>';
      }
      result += '</div>';
    });
    result += '<div class="punctuation brace">}</div>';
    return result;
  }

  parseArray(array) {
    var cnt = array.length;
    var result = '';
    result += '<span class="punctuation hidden">[</span>';
    result += '<span class="array-counter brace punctuation" count="' + cnt + '"></span>';
    for (let i = 0; i < cnt; i++) {
      var elementNo = this.elementsCounter++;

      let lastSibling = (i + 1) === cnt;
      var data = this.parse(array[i], {hasNextSibling: !lastSibling});
      var hasManyChildren = this.elementsCounter - elementNo > 1;
      result += '<div data-element="' + elementNo + '" style="margin-left: 24px" class="node">';
      result += '<span class="array-key-number" index="' + i + '"> &nbsp;</span>';
      result += data;
      if (hasManyChildren) {
        result += '<div data-toggle="' + elementNo + '" class="rootElementToggleButton"></div>';
      }
      result += '</div>';
    }
    result += '<span class="punctuation hidden">],</span>';
    return result;
  }

  parseKey(key) {
    var result = '';
    result += '<span class="key-name">' + key + '</span>';
    return result;
  }
}

self.onmessage = function(e) {
  var parser = new JSONViewer(e.data);
  if (parser.latestError !== null) {
    self.postMessage({
      message: parser.latestError,
      error: true
    });
    return;
  }
  var result = parser.getHTML();
  parser = null;
  self.postMessage({
    message: result,
    error: false
  });
};
